### PR TITLE
`VA_FORMAT_STRING_USES_NEWLINE` handling of text blocks

### LINKED ISSUE #1877: VA_FORMAT_STRING_USES_NEWLINE - False positive in java 17 text block feature
Hi there,
Seems that spotbugs report following error code VA_FORMAT_STRING_USES_NEWLINE on.
```
    // java 17
    public static final String ACCOUNT_QUERY = """
        SELECT id,
               name
        FROM   account
        WHERE  id = '%s'
        """;
    String.format(ACCOUNT_QUERY, "1");
```

AFAIK """ don't accept %n. Guys can you take a look at it? Java 17 is becoming more popular.

### PR DESCRIPTION
**Pull request by gtoison**
The text blocks introduced in Java 15 pose problem when used in a formatter because the line endings are not platform specific and SpotBugs correctly raises a VA_FORMAT_STRING_USES_NEWLINE

Text blocks line endings should be escaped when used in a format (using `%n\`). The bug description should give that solution.

The String.formatted() method (added in Java 15) should also be checked as we are already checking the static version String.format()

This should fix #1877

### CODE REVIEW

**Comment by UngenuineGuineapig** [comment]
I'm clearly not an expert in this field by any means, but imho this is not a fix, but an ugly workaround which goes against the goal of the textblocks. [JEP 378](https://openjdk.org/jeps/378) clearly states that the goal is to _not_ explicitliy write line breaks: "Simplify the task of writing Java programs by making it easy to express strings that span several lines of source code, while avoiding escape sequences in common cases."

Sprinkling %n everywhere does therefore not sound sane. Maybe this is more or less a bug for openjdk and such?

**Comment by gtoison** [comment]
This rule is flagging issues when a string is fed into a formatter, in other cases (merely declaring a string via a text block) it shouldn't complain.

I agree it is ugly but you should write `%n` when using a format and when line endings matter.
In case this doesn't matter the detector can be disabled (or you can suppress the warning)

**Comment by JuditKnoll** [comment]
> I'm clearly not an expert in this field by any means, but imho this is not a fix, but an ugly workaround which goes against the goal of the textblocks. [JEP 378](https://openjdk.org/jeps/378) clearly states that the goal is to _not_ explicitliy write line breaks: "Simplify the task of writing Java programs by making it easy to express strings that span several lines of source code, while avoiding escape sequences in common cases."
> 
> Sprinkling %n everywhere does therefore not sound sane. Maybe this is more or less a bug for openjdk and such?

If you don't want to have line breaks in your string, then you can simply skip the line breaks at the end of the line with `\` like this, you don't need to put `%n`:
```
String value = """
                first line\
                second line\
                """;
```

As far as I understand yes, this is rather a compiler problem, and since SpotBugs works on the compiled code, and I'm not sure whether we can differentiate a simple string literal and the text blocks introduced in Java 15 on the bytecode level (maybe from the line numbers). Also, this is rather about the programmer's intent, whether he/she wants a line break or not.

**Review by JuditKnoll — CHANGES_REQUESTED** [review]

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FormatStringChecker.java [hunk 1]
```diff
@@ -106,19 +106,38 @@ public void sawOpcode(int seen) {
...
 
+            if ("java/lang/String".equals(cl) && "formatted".equals(nm)) {
+                Object stackItem = stack.getStackItem(1).getConstant();
+                if (stackItem instanceof String) {
+                    String string = (String) stackItem;
+                    if (string.indexOf('\n') >= 0) {
+                        bugReporter.reportBug(new BugInstance(this, "VA_FORMAT_STRING_USES_NEWLINE", NORMAL_PRIORITY)
+                                .addClassAndMethod(this).addCalledMethod(this).addString(string)
+                                .describe(StringAnnotation.FORMAT_STRING_ROLE).addSourceLine(this));
+                    }
```
[the remark this answers is not in the document]
Comment (gtoison) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FormatStringChecker.java]: I've made changes to merge the check with the existing one, I think it addresses your remarks
Thank you for reviewing!

**Review by hazendaz — APPROVED** [review]

**Comment by hazendaz** [comment]
Other than concerns on this raised, I think it looks ok.  +1 from me.

File: spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FormatStringChecker.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (hazendaz) [reply spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FormatStringChecker.java]: Think we were good going to merge.

**Pull request merged**